### PR TITLE
[PM-7879, PM-7635] Add server verification for master password to user verification

### PR DESCRIPTION
**Pull request by jlf0dev**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-7879
## 📔 Objective

Adds support for server-side verification of master password to the `UserVerificationService`. Since this was so similar to the logic on the lock component, I have used it there as well.

I also started tests for our user verification service, there's a lot more to be written but this is a good starting point.
## 📸 Screenshots

https://github.com/bitwarden/clients/assets/24985544/65dc50f2-8019-4e64-9f56-52ba9053bc7f

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/auth/services/user-verification/user-verification.service.spec.ts
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/user-verification/user-verification.service.spec.ts:1]: Thank you for these tests! Excellent work and coverage!

**Review by JaredSnider-Bitwarden — COMMENTED** [review]
Really excellent work on this! A few nits and we are good to go!

File: libs/common/src/auth/types/verification.ts [hunk 1]
```diff
@@ -17,3 +19,8 @@ export function verificationHasSecret(
 }
 
 export type ServerSideVerification = OtpVerification | MasterPasswordVerification;
+
+export type MasterPasswordVerificationResponse = {
+  masterKey: MasterKey;
+  policyOptions: MasterPasswordPolicyResponse;
+};
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply libs/common/src/auth/types/verification.ts:26]: This isn't an API response model, so I didn't want to mix the two concepts

File: libs/common/src/auth/abstractions/user-verification/user-verification.service.abstraction.ts [hunk 2]
```diff
@@ -1,19 +1,53 @@
...
   buildRequest: <T extends SecretVerificationRequest>(
     verification: Verification,
     requestClass?: new () => T,
     alreadyHashed?: boolean,
   ) => Promise<T>;
+  /**
+   * Verifies the user using the provided verification data.
+   * PIN or biometrics are verified client-side.
+   * OTP is sent to the server for verification (with no other data)
+   * Master Password attempts to verify client-side first, then server-side if necessary.
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply libs/common/src/auth/abstractions/user-verification/user-verification.service.abstraction.ts:36]: Done!

File: libs/common/src/auth/abstractions/user-verification/user-verification.service.abstraction.ts [hunk 3]
```diff
@@ -22,8 +56,19 @@ export abstract class UserVerificationService {
...
-  ) => Promise<UserVerificationOptions>;
+  /**
+   * Verifies the user using the provided master password.
+   * Attempts to verify client-side first, then server-side if necessary.
+   * IMPORTANT: Will throw an error if the master password is invalid.
+   * @param verification Master Password verification data
+   * @param userId The user to verify
+   * @param email The user's email
+   * @throws Error if the master password is invalid
+   * @returns Master password policy options if the master password is verified on the server, null if verified client-side.
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply libs/common/src/auth/abstractions/user-verification/user-verification.service.abstraction.ts:67]: thanks, forgot to update this.

**Review by JaredSnider-Bitwarden — APPROVED** [review]

**Review by JaredSnider-Bitwarden — APPROVED** [review]

**Pull request merged**